File: sdk/python/loreguard_sdk.py
"""Loreguard SDK for Python game engines.

This SDK helps Python-based games discover and connect to loreguard-client.

Usage:
    from loreguard_sdk import get_base_url, chat

    # Get the URL for loreguard-client
    url = get_base_url()  # e.g., "http://127.0.0.1:52341"

    # Chat with an NPC (streaming)
    async for event in chat("merchant-npc", "What do you have for sale?"):
        if "t" in event:
            print(event["t"], end="", flush=True)
        elif "speech" in event:
            print()  # Done

Requirements:
    pip install httpx
"""

import json
import os
import platform
import subprocess
from pathlib import Path
from typing import Any, AsyncIterator, Optional

try:
    import httpx
except ImportError:
    httpx = None  # Optional dependency


def get_data_dir() -> Path:
    """Get loreguard data directory (matches loreguard-client).

    Returns:
        Path to the loreguard data directory:
        - macOS: ~/Library/Application Support/loreguard
        - Linux: ~/.local/share/loreguard (or $XDG_DATA_HOME/loreguard)
        - Windows: %APPDATA%/loreguard
    """
    if platform.system() == "Windows":
        base = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
    elif platform.system() == "Darwin":
        base = Path.home() / "Library" / "Application Support"
    else:
        base = Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share"))
    return base / "loreguard"


def get_runtime_info() -> Optional[dict]:
    """Load runtime info from loreguard-client.

    Returns:
        Runtime info dict if loreguard-client is running, None otherwise.
        Contains: port, pid, url, started_at, version, backend_connected
    """
    runtime_path = get_data_dir() / "runtime.json"
    if not runtime_path.exists():
        return None
    try:
        with open(runtime_path) as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return None


def get_local_port() -> int:
    """Get the port loreguard-client is running on.

    Returns:
        The port number

    Raises:
        RuntimeError: If loreguard-client is not running
    """
    info = get_runtime_info()
    if info and "port" in info:
        return info["port"]
    raise RuntimeError(
        "loreguard-client not running. "
        "Start it with: loreguard"
    )


def get_base_url() -> str:
    """Get the base URL for loreguard-client API.

    Returns:
        URL like "http://127.0.0.1:52341"

    Raises:
        RuntimeError: If loreguard-client is not running
    """
    port = get_local_port()
    return f"http://127.0.0.1:{port}"


def is_running() -> bool:
    """Check if loreguard-client is running.

    Returns:
        True if running, False otherwise
    """
    info = get_runtime_info()
    if not info:
        return False

    # Verify process is actually running
    pid = info.get("pid")
    if pid:
        try:
            os.kill(pid, 0)
            return True
        except OSError:
            return False
    return False


def get_status_via_cli() -> Optional[dict]:
    """Get status by calling loreguard CLI (fallback method).

    Useful when file access isn't available but subprocess is.

    Returns:
        Status dict if successful, None otherwise
    """
    try:
        result = subprocess.run(
            ["loreguard", "status"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            return json.loads(result.stdout)
    except (subprocess.SubprocessError, json.JSONDecodeError, FileNotFoundError):
        pass
    return None


async def chat(
    character_id: str,
    message: str,
    player_handle: str = "",
    current_context: str = "",
    stream: bool = True,
) -> AsyncIterator[dict[str, Any]]:
    """Chat with an NPC via loreguard-client.

    Args:
        character_id: The NPC's ID
        message: Player's message to the NPC
        player_handle: Player's display name (optional)
        current_context: Game context like "in a dark cave" (optional)
        stream: If True, yields tokens as they arrive. If False, yields final response.

    Yields:
        For streaming: {"t": "token"} for each token, then {"speech": "...", "verified": True, ...}
        For non-streaming: Single dict with complete response

    Raises:
        RuntimeError: If loreguard-client is not running
        ImportError: If httpx is not installed

    Example:
        async for event in chat("merchant", "Hello!"):
            if "t" in event:
                print(event["t"], end="")
            elif "speech" in event:
                print(f"\\nVerified: {event['verified']}")
    """
    if httpx is None:
        raise ImportError("httpx is required for chat. Install with: pip install httpx")

    url = f"{get_base_url()}/api/chat"
    headers = {"Content-Type": "application/json"}

    if stream:
        headers["Accept"] = "text/event-stream"

    body = {
        "character_id": character_id,
        "message": message,
        "player_handle": player_handle,
        "current_context": current_context,
    }

    async with httpx.AsyncClient() as client:
        if stream:
            async with client.stream(
                "POST",
                url,
                headers=headers,
                json=body,
                timeout=120.0,
            ) as response:
                response.raise_for_status()
                async for line in response.aiter_lines():
                    if line.startswith("data: "):
                        try:
                            data = json.loads(line[6:])
                            yield data
                            # Stop after 'done' or 'error' event
                            if "speech" in data or "error" in data:
                                break
                        except json.JSONDecodeError:
                            continue
        else:
            response = await client.post(url, headers=headers, json=body, timeout=120.0)
            response.raise_for_status()
            yield response.json()


async def health_check() -> dict[str, Any]:
    """Check loreguard-client health.

    Returns:
        Health status dict with llm_available, backend_connected, etc.

    Raises:
        RuntimeError: If loreguard-client is not running
        ImportError: If httpx is not installed
    """
    if httpx is None:
        raise ImportError("httpx is required. Install with: pip install httpx")

    url = f"{get_base_url()}/health"
    async with httpx.AsyncClient() as client:
        response = await client.get(url, timeout=5.0)
        response.raise_for_status()
        return response.json()


# Synchronous versions for simpler use cases
def chat_sync(
    character_id: str,
    message: str,
    player_handle: str = "",
    current_context: str = "",
) -> dict[str, Any]:
    """Synchronous chat (non-streaming).

    For async streaming, use the async `chat()` function instead.

    Returns:
        Complete response dict with speech, verified, citations, etc.
    """
    if httpx is None:
        raise ImportError("httpx is required. Install with: pip install httpx")

    url = f"{get_base_url()}/api/chat"
    body = {
        "character_id": character_id,
        "message": message,
        "player_handle": player_handle,
        "current_context": current_context,
    }

    with httpx.Client() as client:
        response = client.post(url, json=body, timeout=120.0)
        response.raise_for_status()
        return response.json()

File: sdk/javascript/loreguard-sdk.js
/**
 * Loreguard SDK for JavaScript / Node.js / Electron
 *
 * This SDK helps JavaScript-based games discover and connect to loreguard-client.
 *
 * Usage (Node.js/Electron):
 *     const { getBaseUrl, chat } = require('./loreguard-sdk');
 *
 *     // Get the URL for loreguard-client
 *     const url = getBaseUrl();  // e.g., "http://127.0.0.1:52341"
 *
 *     // Chat with an NPC (streaming)
 *     for await (const event of chat("merchant-npc", "Hello!")) {
 *       if (event.t) {
 *         process.stdout.write(event.t);
 *       } else if (event.speech) {
 *         console.log("\nVerified:", event.verified);
 *       }
 *     }
 *
 * Usage (Browser - requires bundler):
 *     import { getBaseUrl, chatFetch } from './loreguard-sdk.mjs';
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

/**
 * Get the loreguard data directory path.
 * @returns {string} Path to the loreguard data directory
 */
function getDataDir() {
  switch (process.platform) {
    case 'win32':
      return path.join(process.env.APPDATA || path.join(os.homedir(), 'AppData', 'Roaming'), 'loreguard');
    case 'darwin':
      return path.join(os.homedir(), 'Library', 'Application Support', 'loreguard');
    default:
      // Linux and others
      const xdg = process.env.XDG_DATA_HOME || path.join(os.homedir(), '.local', 'share');
      return path.join(xdg, 'loreguard');
  }
}

/**
 * Get runtime info from loreguard-client.
 * @returns {Object|null} Runtime info if running, null otherwise
 */
function getRuntimeInfo() {
  const runtimePath = path.join(getDataDir(), 'runtime.json');

  if (!fs.existsSync(runtimePath)) {
    return null;
  }

  try {
    const content = fs.readFileSync(runtimePath, 'utf8');
    return JSON.parse(content);
  } catch (e) {
    return null;
  }
}

/**
 * Get the port loreguard-client is running on.
 * @returns {number} Port number
 * @throws {Error} If loreguard-client is not running
 */
function getLocalPort() {
  const info = getRuntimeInfo();
  if (!info || !info.port) {
    throw new Error('loreguard-client not running. Start it with: loreguard');
  }
  return info.port;
}

/**
 * Get the base URL for loreguard-client API.
 * @returns {string} URL like "http://127.0.0.1:52341"
 */
function getBaseUrl() {
  return `http://127.0.0.1:${getLocalPort()}`;
}

function getAuthHeaders() {
  const info = getRuntimeInfo();
  if (!info || !info.api_token) {
    throw new Error('loreguard-client runtime credential is missing; restart loreguard-client');
  }
  return { Authorization: `Bearer ${info.api_token}` };
}

/**
 * Check if loreguard-client is running.
 * @returns {boolean}
 */
function isRunning() {
  try {
    getLocalPort();
    return true;
  } catch {
    return false;
  }
}

/**
 * Chat with an NPC (async generator for streaming).
 * @param {string} characterId - NPC ID
 * @param {string} message - Player's message
 * @param {Object} options - Optional settings
 * @param {string} options.playerHandle - Player's display name
 * @param {string} options.currentContext - Game context
 * @yields {Object} Token events {t: "..."} or completion {speech: "...", verified: true}
 */
async function* chat(characterId, message, options = {}) {
  const url = `${getBaseUrl()}/api/chat`;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'text/event-stream',
      ...getAuthHeaders(),
    },
    body: JSON.stringify({
      character_id: characterId,
      message: message,
      player_handle: options.playerHandle || '',
      player_id: options.playerId || '',
      current_context: options.currentContext || '',
      scenario_id: options.scenarioId || '',
      history: options.history || [],
      chunk_mode: options.chunkMode || '',
      manage_history: options.manageHistory || false,
      max_speech_tokens: options.maxSpeechTokens || 0,
      verbose: options.verbose || false,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop(); // Keep incomplete line

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const data = JSON.parse(line.slice(6));
            yield data;

            // Stop after completion or error
            if (data.speech || data.error) {
              return;
            }
          } catch (e) {
            // Ignore parse errors for incomplete data
          }
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
}

/**
 * Chat with an NPC (simple non-streaming).
 * @param {string} characterId - NPC ID
 * @param {string} message - Player's message
 * @param {Object} options - Optional settings
 * @returns {Promise<Object>} Complete response
 */
async function chatSimple(characterId, message, options = {}) {
  const url = `${getBaseUrl()}/api/chat`;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...getAuthHeaders(),
    },
    body: JSON.stringify({
      character_id: characterId,
      message: message,
      player_handle: options.playerHandle || '',
      player_id: options.playerId || '',
      current_context: options.currentContext || '',
      scenario_id: options.scenarioId || '',
      history: options.history || [],
      chunk_mode: options.chunkMode || '',
      manage_history: options.manageHistory || false,
      max_speech_tokens: options.maxSpeechTokens || 0,
      verbose: options.verbose || false,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Health check for loreguard-client.
 * @returns {Promise<Object>} Health status
 */
async function healthCheck() {
  const url = `${getBaseUrl()}/health`;
  const response = await fetch(url, { headers: getAuthHeaders() });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }

  return response.json();
}

// CommonJS exports
module.exports = {
  getDataDir,
  getRuntimeInfo,
  getLocalPort,
  getBaseUrl,
  getAuthHeaders,
  isRunning,
  chat,
  chatSimple,
  healthCheck,
};
